### PR TITLE
DDP-4896 Country and state not updating properly and DDP-4853 make phone required from activity config

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
Hopefully fixes some of the weirdnesses we have seen recently.
Also tried to implement ensuring that we have phoneNumber if specified in activity component definition from server.
Added comments to PR to explain what I was trying to do.

### CODE REVIEW

File: ddp-workspace/package.json [hunk 1]
```diff
@@ -40,7 +40,7 @@
     "npm-sass": "2.2.1",
     "package": "1.0.1",
     "raw-loader": "0.5.1",
-    "rxjs": "6.5.4",
+    "rxjs": "6.5.5",
```
Comment (MocanaAtBroad) [inline ddp-workspace/package.json:43]: Some bug fixes in rxjs. Might be contributing to some problems.

File: ddp-workspace/projects/ddp-sdk/package.json [hunk 2]
```diff
@@ -27,7 +27,7 @@
     "npm-sass": "2.2.1",
     "package": "1.0.1",
     "raw-loader": "0.5.1",
-    "rxjs": "6.5.4",
+    "rxjs": "6.5.5",
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/package.json:30]: Same library change.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts [hunk 3]
```diff
@@ -64,6 +64,7 @@ import { ConfigurationService } from '../../services/configuration.service';
                                           [readonly]="readonly"
                                           [activityGuid]="activityGuid"
                                           (validStatusChanged)="updateEmbeddedComponentValidationStatus(1, $event)"
+                                          [validationRequested]="validationRequested"
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts:67]: We have been using this input param with data entry components. Will need to do something similar for validation of missing required fields.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 4]
```diff
@@ -65,41 +65,41 @@ interface AddressSuggestion {
...
-                </mat-radio-group>
-            </mat-card-content>
-        </mat-card>
-    </form>`,
+        <p *ngIf="block.titleText" class="ddp-address-embedded__title" [innerHTML]="block.titleText"></p>
+        <p *ngIf="block.subtitleText" class="ddp-address-embedded__subtitle" [innerHTML]="block.subtitleText"></p>
+        <ddp-address-input
+                (valueChanged)="inputComponentAddress$.next($event)"
+                [address]="inputAddress$ | async"
+                [addressErrors]="verifyFieldErrors$ | async"
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:73]: A lot of whitespace changes, but this is the only real change in template. Renamed field for clarity.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 5]
```diff
@@ -135,6 +135,11 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
     this.inputAddress$.next(val);
   }
 
+    @Input()
+    public set validationRequested(val: boolean) {
+        this.validationRequested$.next(val);
+    }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:141]: the validationRequested @Input param converted to subject.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 6]
```diff
@@ -515,17 +522,29 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
         tap(() => busyCounter$.next(-1))
     );
 
+    const canSaveRealAddress = (address: Address | null) =>
+        this.enoughDataToSave(address)  && this.meetsActivityRequirements(address);
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:527]: We have additional criteria to save the address for real.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 7]
```diff
@@ -515,17 +522,29 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
 
+    /* If we get to show validations, we touch controls so we will show validations errors
+        associate with controls
+     */
+    const touchFormOnSubmitWithBadAddressAction$ = this.validationRequested$.pipe(
+        withLatestFrom(currentAddress$),
+        filter(([_, addressToSave]) => !canSaveRealAddress(addressToSave)),
+        tap(() => this.addressInputComponent.touchAllControls())
+    );
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:547]: This makes sure that if phone is required and the field is empty, the validation error associated with control is displayed.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts [hunk 8]
```diff
@@ -544,8 +563,13 @@ export class AddressEmbeddedComponent implements OnDestroy, OnInit {
...
+    const activitityRequirementsMet$: Observable<boolean> = currentAddress$.pipe(
+        map(currentAddress => this.meetsActivityRequirements(currentAddress))
+    );
+
+    const emitValidStatusAction$ = combineLatest([this.formErrorMessages$, this.verifyFieldErrors$, activitityRequirementsMet$]).pipe(
+        map(([formErrors, addressErrors, reqsMet]) =>
+            !formErrors.length && !addressErrors.length && reqsMet),
         distinctUntilChanged(),
         tap(status => this.validStatusChanged.emit(status))
     );
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressEmbedded.component.ts:575]: We emit the validation status of the component. In addition to having form errors, we now take a look as well to the new constraints coming from server (particularly isPhoneRequired)

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.service.ts [hunk 9]
```diff
@@ -132,7 +132,7 @@ export class AddressInputService implements OnDestroy {
       // todo maybe some way to introduce derived updates and make it so state is updated atomically?
       this.addressForm.valueChanges.pipe(
         distinctUntilChanged((x, y) => _.isEqual(x, y)),
-        concatMap((formValue) => {
+        switchMap((formValue) => {
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.service.ts:135]: The  change of "country" from form (that was avoiding switching the state/province field) was stuck right here.
Best explanation I have is an rxjs bug in concatMap: the concatMap function was not executing.
I first tried updating the library and it appeared to fix the problem.
switchMap with either version was working to and is also an acceptable option here.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.service.ts [hunk 10]
```diff
@@ -147,7 +147,7 @@ export class AddressInputService implements OnDestroy {
 
       this.inputAddress$.pipe(
         filter(address => !!address),
-        concatMap((address) => {
+        switchMap((address) => {
           return of(address['country']).pipe(
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/address/addressInput.service.ts:151]: Might as well use switchMap here too.

**Review by fat23cat — APPROVED** [review]
Looks good!
Couldn't reproduce the issues.

**Pull request merged**